File: src/main/java/net/maizegenetics/analysis/data/ExportPlugin.java
/*
 * ExportPlugin.java
 *
 * Created on December 18, 2009
 *
 */
package net.maizegenetics.analysis.data;

import net.maizegenetics.dna.map.PositionList;
import net.maizegenetics.dna.map.PositionListTableReport;
import net.maizegenetics.dna.snp.ExportUtils;
import net.maizegenetics.dna.snp.FilterList;
import net.maizegenetics.dna.snp.GenotypeTable;
import net.maizegenetics.dna.snp.io.FilterJSONUtils;
import net.maizegenetics.dna.snp.io.JSONUtils;
import net.maizegenetics.dna.snp.io.SiteScoresIO;
import net.maizegenetics.phenotype.Phenotype;
import net.maizegenetics.phenotype.PhenotypeUtils;
import net.maizegenetics.plugindef.AbstractPlugin;
import net.maizegenetics.plugindef.DataSet;
import net.maizegenetics.plugindef.PluginParameter;
import net.maizegenetics.taxa.TaxaList;
import net.maizegenetics.taxa.TaxaListTableReport;
import net.maizegenetics.taxa.distance.DistanceMatrix;
import net.maizegenetics.taxa.distance.DistanceMatrixUtils;
import net.maizegenetics.taxa.distance.WriteDistanceMatrix;
import net.maizegenetics.taxa.tree.SimpleTree;
import net.maizegenetics.util.TableReport;
import net.maizegenetics.util.TableReportUtils;
import net.maizegenetics.util.Utils;
import org.apache.log4j.Logger;

import java.io.File;
import java.io.FileWriter;
import java.io.PrintWriter;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;

//import net.maizegenetics.dna.snp.io.FlapjackUtils;
//import net.maizegenetics.taxa.tree.NewickUtils;

/**
 * @author Terry Casstevens
 */
public class ExportPlugin extends AbstractPlugin {

    private static final Logger myLogger = Logger.getLogger(ExportPlugin.class);

    private PluginParameter<String> mySaveFile = new PluginParameter.Builder<>("saveAs", null, String.class)
            .description("Save file as...")
            .outFile()
            .required(true)
            .build();

    private PluginParameter<FileLoadPlugin.TasselFileType> myFileType = new PluginParameter.Builder<>("format", null, FileLoadPlugin.TasselFileType.class)
            .description("Export file format (Default format depends on data being exported)")
            .required(true)
            .objectListSingleSelect()
            .range(FileLoadPlugin.TasselFileType.values())
            .build();

    private PluginParameter<Boolean> myKeepDepth = new PluginParameter.Builder<>("keepDepth", true, Boolean.class)
            .description("Whether to keep depth if format supports depth.")
            .dependentOnParameter(myFileType, new FileLoadPlugin.TasselFileType[]{FileLoadPlugin.TasselFileType.VCF})
            .build();

    private PluginParameter<Boolean> myIncludeTaxaAnnotations = new PluginParameter.Builder<>("includeTaxaAnnotations", true, Boolean.class)
            .description("Whether to include taxa annotations if format supports taxa annotations.")
            .dependentOnParameter(myFileType, new FileLoadPlugin.TasselFileType[]{FileLoadPlugin.TasselFileType.VCF,
                    FileLoadPlugin.TasselFileType.Hapmap,
                    FileLoadPlugin.TasselFileType.HapmapDiploid,
                    FileLoadPlugin.TasselFileType.HapmapLIX})
            .build();

    private PluginParameter<Boolean> myIncludeBranchLengths = new PluginParameter.Builder<>("includeBranchLengths", true, Boolean.class)
            .description("Whether to include branch lengths for Newick formatted files.")
            .dependentOnParameter(myFileType, new FileLoadPlugin.TasselFileType[]{FileLoadPlugin.TasselFileType.Newick})
            .build();

    /**
     * Creates a new instance of ExportPlugin
     */
    public ExportPlugin(boolean isInteractive) {
        super(isInteractive);
    }

    @Override
    protected void preProcessParameters(DataSet input) {

        if (input.getSize() != 1) {
            throw new IllegalArgumentException("Please select only one item.");
        }

        Object data = input.getData(0).getData();
        if (data instanceof GenotypeTable) {
            GenotypeTable genotype = (GenotypeTable) data;
            List<FileLoadPlugin.TasselFileType> temp = new ArrayList<>();
            if (genotype.hasGenotype()) {
                temp.add(FileLoadPlugin.TasselFileType.HaplotypeVCF);
                temp.add(FileLoadPlugin.TasselFileType.Hapmap);
                temp.add(FileLoadPlugin.TasselFileType.HapmapDiploid);
                temp.add(FileLoadPlugin.TasselFileType.VCF);
                temp.add(FileLoadPlugin.TasselFileType.Plink);
                temp.add(FileLoadPlugin.TasselFileType.Phylip_Seq);
                temp.add(FileLoadPlugin.TasselFileType.Phylip_Inter);
                temp.add(FileLoadPlugin.TasselFileType.Table);
            }
            if (genotype.hasDepth()) {
                temp.add(FileLoadPlugin.TasselFileType.Depth);
            }
            if (genotype.hasReferenceProbablity()) {
                temp.add(FileLoadPlugin.TasselFileType.ReferenceProbability);
            }
            myFileType = new PluginParameter<>(myFileType, temp);
        } else if (data instanceof Phenotype) {
            myFileType = new PluginParameter<>(myFileType,
                    Arrays.asList(FileLoadPlugin.TasselFileType.Phenotype,
                            FileLoadPlugin.TasselFileType.PlinkPhenotype));
        } else if (data instanceof FilterList) {
            myFileType = new PluginParameter<>(myFileType,
                    Collections.singletonList(FileLoadPlugin.TasselFileType.Filter));
        } else if (data instanceof DistanceMatrix) {
            myFileType = new PluginParameter<>(myFileType,
                    Arrays.asList(FileLoadPlugin.TasselFileType.SqrMatrix,
                            FileLoadPlugin.TasselFileType.SqrMatrixBin,
                            FileLoadPlugin.TasselFileType.SqrMatrixRaw,
                            FileLoadPlugin.TasselFileType.SqrMatrixDARwinDIS));
        } else if (data instanceof TaxaList) {
            myFileType = new PluginParameter<>(myFileType,
                    Arrays.asList(FileLoadPlugin.TasselFileType.TaxaList,
                            FileLoadPlugin.TasselFileType.Table));
        } else if (data instanceof TaxaListTableReport) {
            myFileType = new PluginParameter<>(myFileType,
                    Arrays.asList(FileLoadPlugin.TasselFileType.TaxaList,
                            FileLoadPlugin.TasselFileType.Table));
        } else if (data instanceof PositionList) {
            myFileType = new PluginParameter<>(myFileType,
                    Arrays.asList(FileLoadPlugin.TasselFileType.PositionList,
                            FileLoadPlugin.TasselFileType.Table));
        } else if (data instanceof PositionListTableReport) {
            myFileType = new PluginParameter<>(myFileType,
                    Arrays.asList(FileLoadPlugin.TasselFileType.PositionList,
                            FileLoadPlugin.TasselFileType.Table));
        } else if (data instanceof TableReport) {
            myFileType = new PluginParameter<>(myFileType,
                    Collections.singletonList(FileLoadPlugin.TasselFileType.Table));
        } else if (data instanceof SimpleTree) {
            myFileType = new PluginParameter<>(myFileType,
                    Arrays.asList(//FileLoadPlugin.TasselFileType.Newick,
                            FileLoadPlugin.TasselFileType.Report));
        } else {
            throw new IllegalStateException("Don't know how to export data type: " + data.getClass().getName());
        }

        if (!isInteractive() && myFileType.isEmpty() && myFileType.hasPossibleValues()) {
            fileType(myFileType.possibleValues().get(0));
        }

    }

    @Override
    public DataSet processData(DataSet input) {

        String filename = null;
        Object data = input.getData(0).getData();
        if (data instanceof GenotypeTable) {
            filename = performFunctionForAlignment((GenotypeTable) data);
        } else if (data instanceof Phenotype) {
            filename = performFunctionForPhenotype((Phenotype) data);
        } else if (data instanceof FilterList) {
            filename = performFunctionForFilter((FilterList) data);
        } else if (data instanceof DistanceMatrix) {
            filename = performFunctionForDistanceMatrix((DistanceMatrix) data);
        } else if (data instanceof TaxaList) {
            filename = performFunctionForTaxaList((TaxaList) data);
        } else if (data instanceof TaxaListTableReport) {
            filename = performFunctionForTaxaList(((TaxaListTableReport) data).getTaxaList());
        } else if (data instanceof PositionList) {
            filename = performFunctionForPositionList((PositionList) data);
        } else if (data instanceof PositionListTableReport) {
            filename = performFunctionForPositionList(((PositionListTableReport) data).getPositionList());
        } else if (data instanceof TableReport) {
            filename = performFunctionForTableReport((TableReport) data);
        } else if (data instanceof SimpleTree) {
            filename = performFunctionForSimpleTree((SimpleTree) data);
        } else {
            throw new IllegalStateException("Don't know how to export data type: " + data.getClass().getName());
        }

        if (filename != null) {
            myLogger.info("performFunction: wrote dataset: " + input.getData(0).getName() + " to file: " + filename);
        }

        return null;

    }

    public String performFunctionForDistanceMatrix(DistanceMatrix input) {

        if (fileType() == FileLoadPlugin.TasselFileType.SqrMatrix) {
            String filename = Utils.addSuffixIfNeeded(saveFile(), ".txt", new String[]{".txt", ".txt.gz"});
            WriteDistanceMatrix.saveDelimitedDistanceMatrix(input, filename);
            return filename;
        } else if (fileType() == FileLoadPlugin.TasselFileType.SqrMatrixRaw) {
            String[] grmFiles = DistanceMatrixUtils.getGRMFilenames(saveFile());
            WriteDistanceMatrix.saveRawMultiBlupMatrix(input, grmFiles[0], grmFiles[3]);
            return grmFiles[3];
        } else if (fileType() == FileLoadPlugin.TasselFileType.SqrMatrixBin) {
            String[] grmFiles = DistanceMatrixUtils.getGRMFilenames(saveFile());
            WriteDistanceMatrix.saveBinMultiBlupMatrix(input, grmFiles[0], grmFiles[1], grmFiles[2]);
            return grmFiles[1];
        } else if (fileType() == FileLoadPlugin.TasselFileType.SqrMatrixDARwinDIS) {
            String filename = Utils.addSuffixIfNeeded(saveFile(), ".dis");
            WriteDistanceMatrix.saveDARwinMatrix(input, filename);
            return filename;
        } else {
            throw new IllegalArgumentException("ExportPlugin: performFunctionForDistanceMatrix: Unknown file type: " + fileType());
        }

    }

    public String performFunctionForTableReport(TableReport input) {
        File theFile = new File(Utils.addSuffixIfNeeded(saveFile(), ".txt"));
        TableReportUtils.saveDelimitedTableReport(input, "\t", theFile);
        return theFile.getAbsolutePath();
    }

    public String performFunctionForFilter(FilterList filter) {
        return FilterJSONUtils.exportFilterToJSON(filter, saveFile());
    }

    public String performFunctionForPhenotype(Phenotype input) {
        String filename = Utils.addSuffixIfNeeded(saveFile(), ".txt");
        if (fileType() == FileLoadPlugin.TasselFileType.Phenotype) {
            PhenotypeUtils.write(input, filename);
        } else if (fileType() == FileLoadPlugin.TasselFileType.PlinkPhenotype) {
            PhenotypeUtils.writePlink(input, filename);
        }
        return new File(filename).getAbsolutePath();
    }

    public String performFunctionForAlignment(GenotypeTable inputAlignment) {

        String resultFile = saveFile();

        if (fileType() == FileLoadPlugin.TasselFileType.ReferenceProbability) {
            resultFile = SiteScoresIO.writeReferenceProbability(inputAlignment, resultFile);
        } else if (fileType() == FileLoadPlugin.TasselFileType.Depth) {
            resultFile = SiteScoresIO.writeDepth(inputAlignment, resultFile);
        } else if (fileType() == FileLoadPlugin.TasselFileType.Hapmap) {
            resultFile = ExportUtils.writeToHapmap(inputAlignment, false, saveFile(), '\t', includeTaxaAnnotations(), this);
        } else if (fileType() == FileLoadPlugin.TasselFileType.HapmapDiploid) {
            resultFile = ExportUtils.writeToHapmap(inputAlignment, true, saveFile(), '\t', includeTaxaAnnotations(), this);
        } else if (fileType() == FileLoadPlugin.TasselFileType.Plink) {
            resultFile = ExportUtils.writeToPlink(inputAlignment, saveFile(), '\t');
            //} else if (fileType() == FileLoadPlugin.TasselFileType.Flapjack) {
            //resultFile = FlapjackUtils.writeToFlapjack(inputAlignment, saveFile(), '\t');
        } else if (fileType() == FileLoadPlugin.TasselFileType.Phylip_Seq) {
            resultFile = Utils.addSuffixIfNeeded(saveFile(), ".phy");
            try (PrintWriter out = new PrintWriter(new FileWriter(resultFile))) {
                ExportUtils.printSequential(inputAlignment, out);
            } catch (Exception e) {
                myLogger.debug(e.getMessage(), e);
                throw new IllegalStateException("ExportPlugin: performFunction: Problem writing file: " + resultFile);
            }
        } else if (fileType() == FileLoadPlugin.TasselFileType.Phylip_Inter) {
            resultFile = Utils.addSuffixIfNeeded(saveFile(), ".phy");
            try (PrintWriter out = new PrintWriter(new FileWriter(resultFile))) {
                ExportUtils.printInterleaved(inputAlignment, out);
            } catch (Exception e) {
                myLogger.debug(e.getMessage(), e);
                throw new IllegalStateException("ExportPlugin: performFunction: Problem writing file: " + resultFile);
            }
        } else if (fileType() == FileLoadPlugin.TasselFileType.Table) {
            resultFile = ExportUtils.saveDelimitedAlignment(inputAlignment, "\t", saveFile());
        } else if (fileType() == FileLoadPlugin.TasselFileType.Serial) {
            resultFile = ExportUtils.writeAlignmentToSerialGZ(inputAlignment, saveFile());
        } else if (fileType() == FileLoadPlugin.TasselFileType.VCF) {
            resultFile = ExportUtils.writeToVCF(inputAlignment, saveFile(), keepDepth(), this);
        } else {
            throw new IllegalStateException("ExportPlugin: performFunction: Unknown Genotype File Format: " + fileType());
        }

        return resultFile;

    }

    public String performFunctionForSimpleTree(SimpleTree input) {

        String resultFile = null;
        if (fileType() == FileLoadPlugin.TasselFileType.Newick) {
            //resultFile = Utils.addSuffixIfNeeded(saveFile(), FileLoadPlugin.FILE_EXT_NEWICK);
            //NewickUtils.write(resultFile, input, includeBranchLengths());
        } else {
            resultFile = Utils.addSuffixIfNeeded(saveFile(), ".txt");
            try (PrintWriter writer = new PrintWriter(resultFile)) {
                input.report(writer);
            } catch (Exception e) {
                myLogger.debug(e.getMessage(), e);
                throw new IllegalStateException("ExportPlugin: performFunctionForReport: Problem writing file: " + resultFile);
            }
        }
        return resultFile;

    }

    public String performFunctionForTaxaList(TaxaList input) {
        if (fileType() == FileLoadPlugin.TasselFileType.TaxaList) {
            return JSONUtils.exportTaxaListToJSON(input, saveFile());
        } else if (fileType() == FileLoadPlugin.TasselFileType.Table) {
            File theFile = new File(Utils.addSuffixIfNeeded(saveFile(), ".txt"));
            TableReportUtils.saveDelimitedTableReport(new TaxaListTableReport(input), "\t", theFile);
            return theFile.getAbsolutePath();
        } else {
            throw new IllegalStateException("ExportPlugin: performFunctionForTaxaList: Can't export TaxaList as: " + fileType());
        }
    }

    public String performFunctionForPositionList(PositionList input) {
        if (fileType() == FileLoadPlugin.TasselFileType.PositionList) {
            return JSONUtils.exportPositionListToJSON(input, saveFile());
        } else if (fileType() == FileLoadPlugin.TasselFileType.Table) {
            File theFile = new File(Utils.addSuffixIfNeeded(saveFile(), ".txt"));
            TableReportUtils.saveDelimitedTableReport(new PositionListTableReport(input), "\t", theFile);
            return theFile.getAbsolutePath();
        } else {
            throw new IllegalStateException("ExportPlugin: performFunctionForPositionList: Can't export PositionList as: " + fileType());
        }
    }

    /**
     * Button name for this plugin to be used in buttons, etc.
     *
     * @return String
     */
    @Override
    public String getButtonName() {
        return "Save As...";
    }

    /**
     * Tool Tip Text for this plugin
     *
     * @return String
     */
    @Override
    public String getToolTipText() {
        return "Save data to files.";
    }

    @Override
    public String pluginUserManualURL() {
        return "https://bitbucket.org/tasseladmin/tassel-5-source/wiki/UserManual/Export/Export";
    }

    /**
     * Save file as...
     *
     * @return Save As
     */
    public String saveFile() {
        return mySaveFile.value();
    }

    /**
     * Save file as...
     *
     * @param value filename
     *
     * @return this plugin
     */
    public ExportPlugin saveFile(String value) {
        mySaveFile = new PluginParameter<>(mySaveFile, value);
        return this;
    }

    @Deprecated
    /**
     * Deprecated use saveFile(value)
     */
    public ExportPlugin setSaveFile(String saveFile) {
        mySaveFile = new PluginParameter<>(mySaveFile, saveFile);
        return this;
    }

    /**
     * Export file format
     *
     * @return Format
     */
    public FileLoadPlugin.TasselFileType fileType() {
        return myFileType.value();
    }

    /**
     * Set Format. Export file format
     *
     * @param value Format
     *
     * @return this plugin
     */
    public ExportPlugin fileType(FileLoadPlugin.TasselFileType value) {
        myFileType = new PluginParameter<>(myFileType, value);
        return this;
    }

    @Deprecated
    /**
     * Deprecated use fileType(value)
     */
    public ExportPlugin setAlignmentFileType(FileLoadPlugin.TasselFileType type) {
        fileType(type);
        return this;
    }

    /**
     * Whether to keep depth if format supports depth.
     *
     * @return Keep Depth
     */
    public Boolean keepDepth() {
        return myKeepDepth.value();
    }

    /**
     * Set Keep Depth. Whether to keep depth if format supports depth.
     *
     * @param value Keep Depth
     *
     * @return this plugin
     */
    public ExportPlugin keepDepth(Boolean value) {
        myKeepDepth = new PluginParameter<>(myKeepDepth, value);
        return this;
    }

    /**
     * Whether to include taxa annotations if format supports taxa annotations.
     *
     * @return Include Taxa Annotations
     */
    public Boolean includeTaxaAnnotations() {
        return myIncludeTaxaAnnotations.value();
    }

    /**
     * Set Include Taxa Annotations. Whether to include taxa annotations if
     * format supports taxa annotations.
     *
     * @param value Include Taxa Annotations
     *
     * @return this plugin
     */
    public ExportPlugin includeTaxaAnnotations(Boolean value) {
        myIncludeTaxaAnnotations = new PluginParameter<>(myIncludeTaxaAnnotations, value);
        return this;
    }

    /**
     * Whether to include branch lengths for Newick formatted
     * files.
     *
     * @return Include Branch Lengths
     */
    public Boolean includeBranchLengths() {
        return myIncludeBranchLengths.value();
    }

    /**
     * Set Include Branch Lengths. Whether to include branch
     * lengths for Newick formatted files.
     *
     * @param value Include Branch Lengths
     *
     * @return this plugin
     */
    public ExportPlugin includeBranchLengths(Boolean value) {
        myIncludeBranchLengths = new PluginParameter<>(myIncludeBranchLengths, value);
        return this;
    }

}

File: src/main/kotlin/net/maizegenetics/analysis/data/ExportPlugin.kt
/*
 * ExportPlugin.java
 *
 * Created on May 1, 2021
 *
 */
package net.maizegenetics.analysis.data

import net.maizegenetics.analysis.data.FileLoadPlugin.TasselFileType
import net.maizegenetics.dna.factor.FeatureTable
import net.maizegenetics.dna.map.PositionList
import net.maizegenetics.dna.map.PositionListTableReport
import net.maizegenetics.dna.snp.ExportUtils
import net.maizegenetics.dna.snp.FilterList
import net.maizegenetics.dna.snp.GenotypeTable
import net.maizegenetics.dna.snp.io.FilterJSONUtils
import net.maizegenetics.dna.snp.io.JSONUtils
import net.maizegenetics.dna.snp.io.SiteScoresIO
import net.maizegenetics.phenotype.Phenotype
import net.maizegenetics.phenotype.PhenotypeUtils
import net.maizegenetics.plugindef.AbstractPlugin
import net.maizegenetics.plugindef.DataSet
import net.maizegenetics.plugindef.Parameter
import net.maizegenetics.plugindef.PluginParameter
import net.maizegenetics.taxa.TaxaList
import net.maizegenetics.taxa.TaxaListTableReport
import net.maizegenetics.taxa.distance.DistanceMatrix
import net.maizegenetics.taxa.distance.DistanceMatrixUtils
import net.maizegenetics.taxa.distance.WriteDistanceMatrix
import net.maizegenetics.taxa.tree.SimpleTree
import net.maizegenetics.util.TableReport
import net.maizegenetics.util.TableReportUtils
import net.maizegenetics.util.Utils
import org.apache.log4j.Logger
import java.io.File
import java.io.FileWriter
import java.io.PrintWriter
import java.util.*

//import net.maizegenetics.dna.snp.io.FlapjackUtils
//import net.maizegenetics.taxa.tree.NewickUtils


/**
 * @author Terry Casstevens
 */
class ExportPlugin(isInteractive: Boolean = false) : AbstractPlugin(isInteractive) {

    private val myLogger = Logger.getLogger(ExportPlugin::class.java)

    private var mySaveFile = PluginParameter.Builder("saveFile", null, String::class.java)
            .description("Save file as...")
            .outFile()
            .required(true)
            .build()
    var saveFile by Parameter<String>()

    private var myFileType = PluginParameter.Builder("fileType", null, TasselFileType::class.java)
            .description("Export file format (Default format depends on data being exported)")
            .required(true)
            .objectListSingleSelect()
            .range(TasselFileType.values())
            .build()
    var fileType by Parameter<TasselFileType?>()

    private var myKeepDepth = PluginParameter.Builder("keepDepth", true, Boolean::class.java)
            .description("Whether to keep depth if format supports depth.")
            .dependentOnParameter(myFileType, arrayOf(TasselFileType.VCF))
            .build()
    var keepDepth by Parameter<Boolean>()

    private var myIncludeTaxaAnnotations = PluginParameter.Builder("includeTaxaAnnotations", true, Boolean::class.java)
            .description("Whether to include taxa annotations if format supports taxa annotations.")
            .dependentOnParameter(myFileType, arrayOf(TasselFileType.VCF,
                    TasselFileType.Hapmap,
                    TasselFileType.HapmapDiploid,
                    TasselFileType.HapmapLIX))
            .build()
    var includeTaxaAnnotations by Parameter<Boolean>()

    private var myIncludeBranchLengths = PluginParameter.Builder("includeBranchLengths", true, Boolean::class.java)
            .description("Whether to include branch lengths for Newick formatted files.")
            .dependentOnParameter(myFileType, arrayOf(TasselFileType.Newick))
            .build()
    var includeBranchLengths by Parameter<Boolean>()


    override fun preProcessParameters(input: DataSet) {
        require(input.size == 1) { "Please select only one item." }
        val data = input.getData(0).data
        myFileType = when (data) {
            is GenotypeTable -> {
                val genotype = data
                val temp: MutableList<TasselFileType?> = ArrayList()
                if (genotype.hasGenotype()) {
                    temp.add(TasselFileType.HaplotypeVCF)
                    temp.add(TasselFileType.Hapmap)
                    temp.add(TasselFileType.HapmapDiploid)
                    temp.add(TasselFileType.VCF)
                    temp.add(TasselFileType.Plink)
                    temp.add(TasselFileType.Phylip_Seq)
                    temp.add(TasselFileType.Phylip_Inter)
                    temp.add(TasselFileType.Table)
                }
                if (genotype.hasDepth()) {
                    temp.add(TasselFileType.Depth)
                }
                if (genotype.hasReferenceProbablity()) {
                    temp.add(TasselFileType.ReferenceProbability)
                }
                PluginParameter(myFileType, temp)
            }
            is FeatureTable -> {
                PluginParameter(myFileType, listOf(TasselFileType.HaplotypeVCF))
            }
            is Phenotype -> {
                PluginParameter(myFileType,
                        listOf(TasselFileType.Phenotype,
                                TasselFileType.PlinkPhenotype))
            }
            is FilterList -> {
                PluginParameter(myFileType, listOf(TasselFileType.Filter))
            }
            is DistanceMatrix -> {
                PluginParameter(myFileType,
                        listOf(TasselFileType.SqrMatrix,
                                TasselFileType.SqrMatrixBin,
                                TasselFileType.SqrMatrixRaw,
                                TasselFileType.SqrMatrixDARwinDIS))
            }
            is TaxaList -> {
                PluginParameter(myFileType,
                        listOf(TasselFileType.TaxaList,
                                TasselFileType.Table))
            }
            is TaxaListTableReport -> {
                PluginParameter(myFileType,
                        listOf(TasselFileType.TaxaList,
                                TasselFileType.Table))
            }
            is PositionList -> {
                PluginParameter(myFileType,
                        listOf(TasselFileType.PositionList,
                                TasselFileType.Table))
            }
            is PositionListTableReport -> {
                PluginParameter(myFileType,
                        listOf(TasselFileType.PositionList,
                                TasselFileType.Table))
            }
            is TableReport -> {
                PluginParameter(myFileType, listOf(TasselFileType.Table))
            }
            is SimpleTree -> {
                PluginParameter(myFileType,
                        listOf( //FileLoadPlugin.TasselFileType.Newick,
                                TasselFileType.Report))
            }
            else -> {
                throw IllegalStateException("Don't know how to export data type: " + data.javaClass.name)
            }
        }
        if (!isInteractive && (fileType == null || fileType == TasselFileType.Unknown) && myFileType.hasPossibleValues()) {
            fileType = myFileType.possibleValues()[0]
        }
    }

    override fun processData(input: DataSet): DataSet? {
        val filename: String?
        val data = input.getData(0).data
        filename = when (data) {
            is GenotypeTable -> performFunctionForAlignment(data)
            is Phenotype -> performFunctionForPhenotype(data)
            is FilterList -> performFunctionForFilter(data)
            is DistanceMatrix -> performFunctionForDistanceMatrix(data)
            is TaxaList -> performFunctionForTaxaList(data)
            is TaxaListTableReport -> performFunctionForTaxaList(data.taxaList)
            is PositionList -> performFunctionForPositionList(data)
            is PositionListTableReport -> performFunctionForPositionList(data.positionList)
            is TableReport -> performFunctionForTableReport(data)
            is SimpleTree -> performFunctionForSimpleTree(data)
            else -> throw IllegalStateException("Don't know how to export data type: " + data.javaClass.name)
        }
        if (filename != null) {
            myLogger.info("performFunction: wrote dataset: ${input.getData(0).name} to file: $filename")
        }
        return null
    }

    fun write(data: Any, filename: String, type: TasselFileType? = null) {
        saveFile = filename
        fileType = type
        performFunction(DataSet.getDataSet(data))
    }

    private fun performFunctionForDistanceMatrix(input: DistanceMatrix): String {
        return if (fileType == TasselFileType.SqrMatrix) {
            val filename = Utils.addSuffixIfNeeded(saveFile, ".txt", arrayOf(".txt", ".txt.gz"))
            WriteDistanceMatrix.saveDelimitedDistanceMatrix(input, filename)
            filename
        } else if (fileType == TasselFileType.SqrMatrixRaw) {
            val grmFiles = DistanceMatrixUtils.getGRMFilenames(saveFile)
            WriteDistanceMatrix.saveRawMultiBlupMatrix(input, grmFiles[0], grmFiles[3])
            grmFiles[3]
        } else if (fileType == TasselFileType.SqrMatrixBin) {
            val grmFiles = DistanceMatrixUtils.getGRMFilenames(saveFile)
            WriteDistanceMatrix.saveBinMultiBlupMatrix(input, grmFiles[0], grmFiles[1], grmFiles[2])
            grmFiles[1]
        } else if (fileType == TasselFileType.SqrMatrixDARwinDIS) {
            val filename = Utils.addSuffixIfNeeded(saveFile, ".dis")
            WriteDistanceMatrix.saveDARwinMatrix(input, filename)
            filename
        } else {
            throw IllegalArgumentException("ExportPlugin: performFunctionForDistanceMatrix: Unknown file type: $fileType")
        }
    }

    private fun performFunctionForTableReport(input: TableReport): String {
        val theFile = File(Utils.addSuffixIfNeeded(saveFile, ".txt"))
        TableReportUtils.saveDelimitedTableReport(input, "\t", theFile)
        return theFile.absolutePath
    }

    private fun performFunctionForFilter(filter: FilterList): String {
        return FilterJSONUtils.exportFilterToJSON(filter, saveFile)
    }

    private fun performFunctionForPhenotype(input: Phenotype): String {
        val filename = Utils.addSuffixIfNeeded(saveFile, ".txt")
        if (fileType == TasselFileType.Phenotype) {
            PhenotypeUtils.write(input, filename)
        } else if (fileType == TasselFileType.PlinkPhenotype) {
            PhenotypeUtils.writePlink(input, filename)
        }
        return File(filename).absolutePath
    }

    private fun performFunctionForFactorTable(table: FeatureTable): String? {

        var resultFile = saveFile
        when (fileType) {
            TasselFileType.HaplotypeVCF -> TODO()
        }

        return resultFile

    }

    private fun performFunctionForAlignment(inputAlignment: GenotypeTable): String? {
        var resultFile = saveFile
        when (fileType) {
            TasselFileType.ReferenceProbability -> resultFile = SiteScoresIO.writeReferenceProbability(inputAlignment, resultFile)
            TasselFileType.Depth -> resultFile = SiteScoresIO.writeDepth(inputAlignment, resultFile)
            TasselFileType.Hapmap -> resultFile = ExportUtils.writeToHapmap(inputAlignment, false, saveFile, '\t', includeTaxaAnnotations, this)
            TasselFileType.HapmapDiploid -> resultFile = ExportUtils.writeToHapmap(inputAlignment, true, saveFile, '\t', includeTaxaAnnotations, this)
            TasselFileType.Plink -> resultFile = ExportUtils.writeToPlink(inputAlignment, saveFile, '\t')
            //TasselFileType.Flapjack -> resultFile = FlapjackUtils.writeToFlapjack(inputAlignment, saveFile(), '\t')
            TasselFileType.Phylip_Seq -> {
                resultFile = Utils.addSuffixIfNeeded(saveFile, ".phy")
                try {
                    PrintWriter(FileWriter(resultFile)).use { out -> ExportUtils.printSequential(inputAlignment, out) }
                } catch (e: Exception) {
                    myLogger.debug(e.message, e)
                    throw IllegalStateException("ExportPlugin: performFunction: Problem writing file: $resultFile")
                }
            }
            TasselFileType.Phylip_Inter -> {
                resultFile = Utils.addSuffixIfNeeded(saveFile, ".phy")
                try {
                    PrintWriter(FileWriter(resultFile)).use { out -> ExportUtils.printInterleaved(inputAlignment, out) }
                } catch (e: Exception) {
                    myLogger.debug(e.message, e)
                    throw IllegalStateException("ExportPlugin: performFunction: Problem writing file: $resultFile")
                }
            }
            TasselFileType.Table -> resultFile = ExportUtils.saveDelimitedAlignment(inputAlignment, "\t", saveFile)
            TasselFileType.Serial -> resultFile = ExportUtils.writeAlignmentToSerialGZ(inputAlignment, saveFile)
            TasselFileType.VCF -> resultFile = ExportUtils.writeToVCF(inputAlignment, saveFile, keepDepth, this)
            else -> throw IllegalStateException("ExportPlugin: performFunction: Unknown Genotype File Format: $fileType")
        }
        return resultFile
    }

    private fun performFunctionForSimpleTree(input: SimpleTree): String? {
        var resultFile: String? = null
        if (fileType == TasselFileType.Newick) {
            //resultFile = Utils.addSuffixIfNeeded(saveFile(), FileLoadPlugin.FILE_EXT_NEWICK);
            //NewickUtils.write(resultFile, input, includeBranchLengths());
        } else {
            resultFile = Utils.addSuffixIfNeeded(saveFile, ".txt")
            try {
                PrintWriter(resultFile).use { writer -> input.report(writer) }
            } catch (e: Exception) {
                myLogger.debug(e.message, e)
                throw IllegalStateException("ExportPlugin: performFunctionForReport: Problem writing file: $resultFile")
            }
        }
        return resultFile
    }

    private fun performFunctionForTaxaList(input: TaxaList): String {
        return if (fileType == TasselFileType.TaxaList) {
            JSONUtils.exportTaxaListToJSON(input, saveFile)
        } else if (fileType == TasselFileType.Table) {
            val theFile = File(Utils.addSuffixIfNeeded(saveFile, ".txt"))
            TableReportUtils.saveDelimitedTableReport(TaxaListTableReport(input), "\t", theFile)
            theFile.absolutePath
        } else {
            throw IllegalStateException("ExportPlugin: performFunctionForTaxaList: Can't export TaxaList as: " + fileType)
        }
    }

    private fun performFunctionForPositionList(input: PositionList): String {
        return if (fileType == TasselFileType.PositionList) {
            JSONUtils.exportPositionListToJSON(input, saveFile)
        } else if (fileType == TasselFileType.Table) {
            val theFile = File(Utils.addSuffixIfNeeded(saveFile, ".txt"))
            TableReportUtils.saveDelimitedTableReport(PositionListTableReport(input), "\t", theFile)
            theFile.absolutePath
        } else {
            throw IllegalStateException("ExportPlugin: performFunctionForPositionList: Can't export PositionList as: " + fileType)
        }
    }

    /**
     * Button name for this plugin to be used in buttons, etc.
     *
     * @return String
     */
    override fun getButtonName(): String {
        return "Save As..."
    }

    /**
     * Tool Tip Text for this plugin
     *
     * @return String
     */
    override fun getToolTipText(): String {
        return "Save data to files."
    }

    override fun pluginUserManualURL(): String {
        return "https://bitbucket.org/tasseladmin/tassel-5-source/wiki/UserManual/Export/Export"
    }

}